Updating ads-deploy to latest pytmc and ads-ioc, upgraded dependencie…

yannS2016 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
## Description
<!--- Describe your changes in detail -->

- Updating ads-deploy to the latest pytmc and ads-ioc, upgraded dependencies to the current pcds conda
- Added a bash script to fix the issue with Python3 missing in the ads-deploy env while building the ioc, and pull the latest tag after installing the env, the newest release is now used as the default ads-ioc

**Note**: adjust the SSH key and the ssh-agent for your shell. On Git-Bash run: `cmd.exe /c setup_conda_on_windows.cmd`
## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
https://jira.slac.stanford.edu/browse/ECS-9455
## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
https://jira.slac.stanford.edu/browse/ECS-9455
## Where Has This Been Documented?
<!--  Include where the changes made have been documented. -->
<!--  This can simply be  a comment in the code or updating a docstring -->
https://jira.slac.stanford.edu/browse/ECS-9455
<!--
## Screenshots (if appropriate):
-->
<img width="3402" height="1915" alt="build-ioc-ads-deploy-2 12 0" src="https://github.com/user-attachments/assets/3223780d-ffc5-4951-b60d-ca0f7c79f7ca" />

yannS2016 committed:
Updating ads-deploy to latest pytmc and ads-ioc, upgraded dependencies to currents pcds conda
added a bash script to fix issue with python3 missing in ads-deploy env, pull latest tag after installing the env and using it as the default ads-deploy.

KaushikMalapati commented on requirements.txt:
```diff
@@ -1,4 +1,4 @@
-pytmc>=2.11.0
-typhos>=1.0
+pytmc>=2.19.0
+typhos>=4.2.0
```
Why is typhos changing?

yannS2016 replied:
just mirroring the current status of PCDs Conda. that's the version being used

tangkong commented on requirements.txt:
```diff
@@ -1,4 +1,4 @@
+pytmc>=2.19.0
+typhos>=4.2.0
 jinja2
 versioneer
```
This is probably overdue for a transition from versioneer to setuptools_scm

yannS2016 replied:
It should be indeed. We could look at it on a subsequent PR.

KaushikMalapati commented on ads_deploy/windows/build_ioc.cmd:
```diff
@@ -12,7 +12,7 @@ IF %AdsDeployUseDocker% EQU 1 (
     %RunDocker% %DockerImage% "find '%IocMountPath%/iocBoot' -type d -maxdepth 1 -name 'ioc*' -exec make -C {} \;"
 ) ELSE (
-    C:\miniconda\envs\ads-deploy-2.9.1\Library\usr\bin\bash.exe --login -c "find '%SolutionDir:\=/%/iocBoot' -type d -maxdepth 1 -name 'ioc*' -exec bash --login %AdsDeployWindowsScripts:\=/%/build.sh '{}' '%WINDOWS_ADS_IOC_TOP:\=/%/' \;"
+    bash.exe --login -c "find '%SolutionDir:\=/%/iocBoot' -type d -maxdepth 1 -name 'ioc*' -exec bash --login %AdsDeployWindowsScripts:\=/%/build.sh '{}' '%WINDOWS_ADS_IOC_TOP:\=/%/' \;"
```
Why not use the absolute path?

yannS2016 replied:
No particular reason to use a hardcoded path here, the "bash" is supposed to be part of the activated conda env anyway.

KaushikMalapati commented on setup_conda_on_windows.cmd:
```diff
@@ -58,8 +58,10 @@ IF %ERRORLEVEL% NEQ 0 (
 
-bash -c "git clone https://github.com/pcdshub/ads-ioc/ '%WINDOWS_ADS_IOC_TOP%'/master || (cd '%WINDOWS_ADS_IOC_TOP%'/master && git pull origin master)"
-bash -c "git clone --single-branch --branch R0.2.4 '%WINDOWS_ADS_IOC_TOP%'/master '%WINDOWS_ADS_IOC_TOP%'/R0.2.4"
+@echo off
```
Why is this needed if line 1 does `@ECHO OFF`?

yannS2016 replied:
Probably a duplicate.

KaushikMalapati commented on setup_conda_on_windows.cmd:
```diff
@@ -58,8 +58,10 @@ IF %ERRORLEVEL% NEQ 0 (
+@echo off
+REM ssh-agent must be setup on shell invoking this line.
+REM install ads-deploy from git bash( if you setup you ssh agent there):  cmd.exe /c setup_conda_on_windows.cmd
+bash --login -c "$(pwd)/ads_deploy/windows/add_on.sh '%WINDOWS_ADS_IOC_TOP%'"
```
Why is a login shell needed?

KaushikMalapati commented:
@yannS2016 can you explain what you mean by these comments?
- Note: adjust the SSH key and the ssh-agent for your shell. On Git-Bash run: cmd.exe /c setup_conda_on_windows.cmd
- install ads-deploy from git bash( if you setup you ssh agent there):  cmd.exe /c setup_conda_on_windows.cmd

KaushikMalapati commented on ads_deploy/windows/post_install.sh:
```diff
@@ -0,0 +1,101 @@
+#!/bin/bash
+
+set -e
```
You can use a trap instead and print a message letting the user know that this script failed
https://www.linuxjournal.com/content/bash-trap-command

yannS2016 replied:
This could be improved at a later stage.

KaushikMalapati commented on ads_deploy/windows/post_install.sh:
```diff
@@ -0,0 +1,101 @@
+#!/bin/bash
+
+set -e
+export LC_ALL=C
```
Why is this needed?

yannS2016 replied:
sets the locale for all locale-dependent operations in a deterministic and OS-independent fashion.

KaushikMalapati commented on ads_deploy/windows/post_install.sh:
```diff
@@ -0,0 +1,101 @@
+set -e
+export LC_ALL=C
+
+die() { echo "ERROR: $*" >&2; exit 1; }
```
Can't this use $1 since it only takes one argument?

yannS2016 replied:
indeed. It could be.

ZLLentz commented on ads_deploy/windows/post_install.sh:
```diff
@@ -0,0 +1,101 @@
+fi
+
+# 3. Get latest release tag from GitHub
+LATEST_TAG="$(curl -s https://api.github.com/repos/pcdshub/ioc-common-ads-ioc/releases/latest | grep -o '"tag_name": *"[^"]*"' | sed 's/"tag_name": *"//;s/"//g')"
```
This likely breaks if we make ads-ioc an internal repo

yannS2016 replied:
Yeah, most probably. Any workaround you may be looking at?

ZLLentz replied:
Not a workaround- we'll need to authenticate in this case. My suggestion is to document all of these review suggestions you're not going to handle in this PR as github issues.

ZLLentz commented on ads_deploy/windows/post_install.sh:
What does this script do? Can it be documented in the script?

yannS2016 replied:
I'll be adding some documentation

yannS2016 replied:
@ZLLentz added a docstring to the add-on script.

ZLLentz replied:
It's missing the most important detail: when/why do I call this script?
I appreciate some of the other details but this is the bit that matters.

"This script is used to (upgrade? update? install?) (ads-deploy? ads-ioc?) on (any computer? any windows machine?) because (doing this manually is hard?). You would call this script when (?)" is the main thing I was hoping to learn- rather than a laundry list of steps without context.

I guess your description from the PR is closer to what I was looking for?
> fix the issue with Python3 missing in the ads-deploy env while building the ioc, and pull the latest tag after installing the env

ZLLentz replied:
I'm also super confused by the name of the script but it's not so important. What am I adding on?

yannS2016 replied:
renamed it as `post_install`

ZLLentz replied:
OK, the name post_install makes it clear enough: thank you! (now the next person who comes across the script will have some understanding of what it is for)

yannS2016 commented:
> @yannS2016 can you explain what you mean by these comments?
> 
> * Note: adjust the SSH key and the ssh-agent for your shell. On Git-Bash run: cmd.exe /c setup_conda_on_windows.cmd
> * install ads-deploy from git bash( if you setup you ssh agent there):  cmd.exe /c setup_conda_on_windows.cmd

SSH must be appropriately set up for the shell running the script. The recommendation is git-bash. wasn't getting expected results from CMD.

yannS2016 committed:
Added a doc string to add_on scripts and fixed duplicate echo in conda_setup_windows

yannS2016 commented:
I addressed most comments from the initial review. What would be an issue was noted by @ZLLentz regarding ads-ioc changing to an internal repo. The script functions as intended.

ZLLentz commented on ads_deploy/windows/add_on.sh:
```diff
@@ -0,0 +1,121 @@
+
+SCRIPT_PATH="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+WINDOWS_ADS_IOC_TOP="$1"
+
+source /c/miniconda/etc/profile.d/conda.sh
+
+source "${SCRIPT_PATH}/conda_config.sh"
+conda activate $ADS_DEPLOY_CONDA_ENV
+
+CONDA_ENV_PATH="${CONDA_PREFIX:-}"
+if [[ -z "$CONDA_ENV_PATH" ]]; then
+    die "CONDA_PREFIX is not set. Please activate your conda env first!"
+fi
```
I don't understand this error message. This script appears to activate a conda environment on line 31, what specifically does the user need to do here?

yannS2016 replied:
Pushed a new update to optimize these checks. Ensure the environment is activated by checking the "CONDA_PREFIX" variable, which is set after activation.

ZLLentz commented on ads_deploy/windows/add_on.sh:
```diff
@@ -0,0 +1,121 @@
+mv "$TEMP_CONFIG" "$CONFIG_FILE"
+echo "Updated $CONFIG_FILE with WINDOWS_ADS_IOC_TOP=$LATEST_IOC_TOP"
+
+# echo "All steps complete."
```
Lingering comment, should be removed

ZLLentz commented on ads_deploy/windows/post_install.sh:
```diff
@@ -0,0 +1,121 @@
+##############################################################################
+# add_on.sh
+#
+# - Ensures the active conda environment's Python executable is available as "python3"
```
Why is this important?

yannS2016 replied:
Building the plc ioc under the activating env looks for Python, but the env only has Python by default. This would error the build process.

ZLLentz commented on ads_deploy/windows/conda_config.cmd:
I'm a bit confused- is this one of the files updated by `add_on.sh`? I misunderstood and assumed that this script was only updating deployments on a windows machine, but this seems to be a file in the repo?

yannS2016 replied:
yeah, add_on script update the ads-ioc env vars to point to the latest tag path. My bad for the poorly named script.

yannS2016 committed:
Renamed add-on script to post-install optimize implementation

ZLLentz reviewed (approved):
OK, my confusion is gone and I think we're past the important threshold where someone else could use these tools without needing extra help. Thank you for the good work Yann!
